### PR TITLE
Some more tidying based on remaining tickets and PR comments.

### PR DESCRIPTION
**Pull request by jdunkerley**
### Pull Request Description

- Moved `to_default_visualization_data` to `Standard.Visualization`.
- Remove the use of `is_a` in favour of case statements.
- Stop exporting Standard.Base.Error.Common.
  - Separate errors to own files.
  - Change constructors to be called `Error`.
- Rename `Caught_Panic.Caught_Panic_Data` -> `Caught_Panic.Panic`.
- Rename `Project_Description.Project_Description_Data` ->`Project_Description.Value`
- Rename `Regex_Matcher.Regex_Matcher_Data` -> `Regex_Matcher.Value` (can't come up with anything better!).
- Rename `Range.Value` -> `Range.Between`.
- Rename `Interval.Value` -> `Interval.Between`.
- Rename `Column.Column_Data` -> `Column.Value`.
- Rename `Table.Table_Data` -> `Table.Value`.
- Align all the Error types in Table.
- Removed GEO Json bits from Table.
  - `Json.to_table` doesn't have the GEO bits anymore.
  - Added `Json.geo_json_to_table` to add the functions back in.

### Important Notes

No more exports from anywhere but Main! 
No more `_Data` constructors!

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide build` and `./run ide watch`.

### CODE REVIEW

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/UniquelyConstructibleBuiltin.java [hunk 1]
```diff
@@ -15,7 +15,11 @@ public final AtomConstructor getUniqueConstructor() {
 
   @Override
   protected final List<Cons> getDeclaredConstructors() {
-    return List.of(new Cons(getName() + "_Data", getConstructorParamNames()));
+    return List.of(new Cons(getConstructorName(), getConstructorParamNames()));
+  }
+
+  protected String getConstructorName() {
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/UniquelyConstructibleBuiltin.java:21]: Useful extension.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Statistics.enso [hunk 2]
```diff
@@ -137,7 +139,12 @@ type Statistic
        - statistics: Set of statistics to calculate.
     compute_bulk : Vector -> Vector Statistic -> Vector Any
     compute_bulk data statistics=[Statistic.Count, Statistic.Sum] =
-        count_min_max = statistics.any s->((s.is_a Statistic.Count) || (s.is_a Statistic.Minimum) || (s.is_a Statistic.Maximum))
+        count_min_max = statistics.any s->
```
Comment (JaroslavTulach) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Statistics.enso:142]: Thanks. I think the use of `is_a` here was a mistake. Checking for constructor seems more correct. This change shall fix https://github.com/enso-org/enso/pull/3949#discussion_r1040058599 failures.

**Review by JaroslavTulach — APPROVED** [review]
Checked Java & Scala changes. They are OK. Glanced thru the `.enso` files. Seem OK too.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex/Engine/Default.enso [hunk 3]
```diff
@@ -829,13 +829,9 @@ type Match
...
 handle_error : Any -> (Text | Integer) -> Any
 handle_error error id = case error of
-    Polyglot_Error.Polyglot_Error_Data err ->
-        is_ioob = err.is_a IndexOutOfBoundsException
-        is_iae = err.is_a IllegalArgumentException
-        maps_to_no_such_group = is_ioob || is_iae
-
-        if maps_to_no_such_group.not then err else
-            No_Such_Group.Error id
+    Polyglot_Error.Error (_ : IndexOutOfBoundsException) -> No_Such_Group.Error id
```
Comment (hubertp) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex/Engine/Default.enso]: knowing how this is desugared I would probably rewrite this into a nested  case :p

File: distribution/lib/Standard/Visualization/0.0.0-dev/src/Main.enso [hunk 4]
```diff
@@ -1,8 +1,10 @@
...
 import project.File_Upload
-import project.Id
+import project.Id.Id
 import project.Preprocessor
+import project.Helpers
 
 from project.File_Upload export file_uploading
-export project.Id
+export project.Id.Id
+export project.Helpers
```
Comment (hubertp) [inline distribution/lib/Standard/Visualization/0.0.0-dev/src/Main.enso]:
```suggestion
export project.Helpers

```

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/text/util/TypeToDisplayTextNode.java [hunk 5]
```diff
@@ -52,11 +50,12 @@ String doDisplay(
     } else if (TypesGen.isFunction(value)) {
       return "Function";
     } else if (value instanceof Atom atom) {
-      var cons = atom.getConstructor();
-      if (cons.getName().equals("Value")) {
-        return cons.getType().getName() + "." + cons.getName();
+      var constructor = atom.getConstructor();
+      var consName = constructor.getName();
+      if (consName.equals("Value") || consName.equals("Error")) {
```
Comment (hubertp) [inline engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/text/util/TypeToDisplayTextNode.java]: Shoud we maybe push this into Constructor? This starts to look like some hardcoded magic

**Review by hubertp — APPROVED** [review]

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data.enso [hunk 6]
```diff
@@ -18,7 +18,7 @@ from project.Error.Common import Unsupported_File_Type, File_Error
    - format: A `File_Format` object used to read file into memory.
      If `Auto_Detect` is specified; the provided file determines the specific
      type and configures it appropriately. If there is no matching type then
-     a `Unsupported_File_Type` error is returned.
+     a `File_Format.Unsupported_Type` error is returned.
```
Comment (radeusgd) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data.enso]:
```suggestion
     a `File_Error.Unsupported_Type` error is returned.
```

File: distribution/lib/Standard/Base/0.0.0-dev/src/Error/Common.enso [hunk 7]
```diff
@@ -244,36 +148,24 @@ from project.Error.Common.Unsupported_Argument_Types export all
...
-    Invalid_Conversion_Target_Error_Data target
-
-## TODO Dubious constructor export
-from project.Error.Common.No_Such_Conversion_Error import all
-from project.Error.Common.No_Such_Conversion_Error export all
+type Invalid_Conversion_Target
+    Error target
 
 ## An error that occurs when the conversion from one type to another does not exist
 ## FIXME: please check
```
Comment (radeusgd) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Error/Common.enso]: Can we address that check by any chance?

File: distribution/lib/Standard/Base/0.0.0-dev/src/Error/Illegal_Argument.enso [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+    Error message cause=Nothing
+
+    ## PRIVATE
+
+       Provides a human-readable representation of the encoding error.
+    to_display_text : Text
+    to_display_text self = "Illegal Argument: " + self.message
+
+    ## PRIVATE
+       Capture a Java IllegalArgumentException and rethrow
```
Comment (radeusgd) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Error/Illegal_Argument.enso]:
```suggestion
       Capture a Java `IllegalArgumentException` and convert it to an Enso dataflow error - `Illegal_Argument_Error`.
```

File: distribution/lib/Standard/Base/0.0.0-dev/src/Error/Incomparable_Values.enso [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+polyglot java import java.lang.ClassCastException
+
+## UNSTABLE
+
+   An error indicating that some elements are incomparable.
+
+type Incomparable_Values
+    ## ADVANCED
+       Catches possible errors from comparing values and throws an
+       Incomparable_Values if any occur.
```
Comment (radeusgd) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Error/Incomparable_Values.enso]:
```suggestion
       `Incomparable_Values` if any occur.
```

File: distribution/lib/Standard/Base/0.0.0-dev/src/Error/Time_Error.enso [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+
+type Time_Error
+
+    ## UNSTABLE
+
+       An error produced while working with time- and date-related methods.
+
+       Arguments:
+       - error_message: The message for the error.
+    Error error_message
```
Comment (radeusgd) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Error/Time_Error.enso]: I don't like that we are inconsistent with doc comments - some are attached to the _type_ while some are attached to a _constructor_.

IIRC the latter is needed to make these docs properly show up in the searcher (unless something has changed), while the former looks much better in text mode.

Regardless of the choice, we should at least be consistent here.

File: distribution/lib/Standard/Base/0.0.0-dev/src/System/File.enso [hunk 11]
```diff
@@ -161,7 +163,7 @@ type File
        - format: A `File_Format` object used to read file into memory.
          If `Auto_Detect` is specified; the provided file determines the specific
          type and configures it appropriately. If there is no matching type then
-         a `Unsupported_File_Type` error is returned.
+         a `File_Format.Unsupported_Type` error is returned.
```
Comment (radeusgd) [inline distribution/lib/Standard/Base/0.0.0-dev/src/System/File.enso]:
```suggestion
         a `File_Error.Unsupported_Type` error is returned.
```

File: distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso [hunk 12]
```diff
@@ -809,7 +814,12 @@ type Table
                 agg = p.second
                 new_name = p.first
                 Aggregate_Helper.make_aggregate_column self agg new_name . catch
-            partitioned = results.partition (_.is_a Internal_Column.Value)
+
+            is_a_column v = case v of
+                _ : Internal_Column -> True
+                _ -> False
+            partitioned = results.partition is_a_column
```
Comment (radeusgd) [inline distribution/lib/Standard/Database/0.0.0-dev/src/Data/Table.enso]: Given the pending PR #3949 , I'm not sure if this change is desirable? Having `is_a` here seems like a much better choice and after the fix its performance should be on-par with pattern matching. Moreover, this is not a critical path anyway, so it souldn't really matter.
The new way is less readable - if it were for a really important performance gain it would be ok - but given the pending PR I'm not convinced we need this change.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex/Engine/Default.enso [hunk 13]
```diff
@@ -132,9 +132,9 @@ type Default_Engine
             Java_Pattern.compile (unicode_regex.transform expression) options_bitmask
 
         internal_pattern = maybe_java_pattern.map_error case _ of
-            Polyglot_Error.Polyglot_Error_Data err ->
-                if err.is_a PatternSyntaxException . not then err else
-                    Syntax_Error_Data ("The regex could not be compiled: " + err.getMessage)
+            Polyglot_Error.Error (err : PatternSyntaxException) ->
+                Syntax_Error.Error ("The regex could not be compiled: " + err.getMessage)
+            Polyglot_Error.Error err -> err
```
Comment (radeusgd) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex/Engine/Default.enso]: I guess @hubertp's comment from below applies here too, right?

Btw. @hubertp can you explain to us what this desugars to and why it should be avoided in this case?

File: distribution/lib/Standard/Base/0.0.0-dev/src/Error/Time_Error.enso (hunk 10 above)
Comment (radeusgd) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Error/Time_Error.enso]: Thank you

File: distribution/lib/Standard/Visualization/0.0.0-dev/src/Preprocessor.enso [hunk 14]
```diff
@@ -12,5 +13,7 @@ error_preprocessor x =
     ok = '{ message: ""}'
     result = x.map_error err->
         message = err.to_display_text
-        '{ "kind": "Dataflow", "message": ' + message.to_json.to_text + '}'
+        stack_trace = x.get_stack_trace_text.if_nothing "" . split '\n'
+        full_message = message + if stack_trace.length > 1 then " (" + stack_trace.at 1 . trim +")" else ""
```
Comment (radeusgd) [inline distribution/lib/Standard/Visualization/0.0.0-dev/src/Preprocessor.enso:17]: Why `at 1` instead of `at 0`?

Also, what about including additional property - `stack_trace` with the whole trace? Adding a field should be compatible with the current vis, but a new vis could then easily use it to display better traces.

Comment (jdunkerley) [reply distribution/lib/Standard/Visualization/0.0.0-dev/src/Preprocessor.enso:17]: `at 1` skips the `Error.throw` attached to the Error.
This was just a quick win.

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/error/ArithmeticError.java [hunk 15]
```diff
@@ -7,6 +7,11 @@
 
 @BuiltinType
 public class ArithmeticError extends UniquelyConstructibleBuiltin {
+  @Override
+  protected String getConstructorName() {
+    return "Error";
+  }
```
Comment (radeusgd) [inline engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/error/ArithmeticError.java:13]: I'm wondering if it would be worth creating a subclass with that override - then we could just override `ErrorBuiltin` or something like that to get this implementation without repetition?

Not sure if worth it, but may simplify a little bit.

File: engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/error/displaytext/InexhaustivePatternMatchToDisplayTextNode.java [hunk 16]
```diff
@@ -9,10 +9,10 @@
...
-public abstract class InexhaustivePatternMatchErrorToDisplayTextNode extends Node {
-  static InexhaustivePatternMatchErrorToDisplayTextNode build() {
-    return InexhaustivePatternMatchErrorToDisplayTextNodeGen.create();
+@BuiltinMethod(type = "Inexhaustive_Pattern_Match", name = "to_display_text")
+public abstract class InexhaustivePatternMatchToDisplayTextNode extends Node {
+  static InexhaustivePatternMatchToDisplayTextNode build() {
+    return InexhaustivePatternMatchToDisplayTextNodeGen.create();
   }
 
   abstract Text execute(Object self);
```
Comment (radeusgd) [inline engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/error/displaytext/InexhaustivePatternMatchToDisplayTextNode.java:18]: I wonder if we could extend our `Builtin` class to auto-generate these just from a `toText` method? It is a bit tedious having to update these each time (although I guess the refactor tool deals with these OK so it's mostly tedious to review :wink: )

Comment (jdunkerley) [reply engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/error/displaytext/InexhaustivePatternMatchToDisplayTextNode.java:18]: I raised with @JaroslavTulach as to whether we even need these anymore and shouldn't just go with methods in the enso file. Not one for this PR though.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex/Engine/Default.enso (hunk 13 above)
Comment (jdunkerley) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex/Engine/Default.enso]: Removed this usage as well.

File: test/Tests/src/Data/Array_Spec.enso [hunk 17]
```diff
@@ -47,21 +39,4 @@ spec =
...
-
-    Test.group "Polyglot Arrays" <|
-        test_arrays make_java_array
-
-        Test.specify "should be able to be converted to a visualization rep" pending="`to_default_visualization_data` does not work for polyglot arrays" <|
-            arr = make_java_array (Vector.fill 1000 0)
-            text = arr.to_default_visualization_data
-            json = Json.parse text
-            as_vec = json.into (Vector.fill 1 Number)
-            as_vec.should_equal <| Vector.fill 100 0
```
Comment (radeusgd) [inline test/Tests/src/Data/Array_Spec.enso:65]: Why is this removed?

Comment (jdunkerley) [reply test/Tests/src/Data/Array_Spec.enso:65]: Moved to the Visualisation Spec.

Comment (radeusgd) [reply test/Tests/src/Data/Array_Spec.enso:65]: Ahh, sorry, I was looking for it but couldn't find it - the last few files are not loaded automatically and I didn't reach there yet - was looking for some `Json_Spec` and gave up.

Looks all good then, sorry.

File: test/Tests/src/Data/Time/Date_Spec.enso [hunk 18]
```diff
@@ -143,8 +144,10 @@ spec_with name create_new_date parse_date =
             (create_new_date 100 1 2).day    . should_equal 2
 
         Test.specify "Gregorian calendar related functionality should produce warning before epoch start" <|
-            expect_warning value =
-                (Warning.get_all value . map .value . exists (_.is_a Time_Error.Error)) . should_be_true
+            is_time_error v = case v of
+                _ : Time_Error -> True
+                _ -> False
+            expect_warning value = (Warning.get_all value . map .value . exists is_time_error) . should_be_true
```
Comment (radeusgd) [inline test/Tests/src/Data/Time/Date_Spec.enso:150]: Again, I think I'd revert this change in the light of #3949

Comment (radeusgd) [reply test/Tests/src/Data/Time/Date_Spec.enso:150]: Especially as it's in tests where this miniscule performance difference doesn't really matter, readability is much more precious here I think.

File: test/Tests/src/Data/Time/Period_Spec.enso [hunk 19]
```diff
@@ -27,9 +29,9 @@ spec =
             (Period.between (Date.new 2022 10 19) (Date.new 2022 11 01)) . should_equal (Period.new days=13)
 
         Test.specify "should not compare between two periods" <|
-            ((Period.new days=10) > (Period.new days=1)) . should_fail_with Incomparable_Values_Error
-            ((Period.new years=10) > (Period.new days=1)) . should_fail_with Incomparable_Values_Error
-            ((Period.new years=10 months=3) > (Period.new months=5)) . should_fail_with Incomparable_Values_Error
+            ((Period.new days=10) > (Period.new days=1)) . should_fail_with Incomparable_Values
+            ((Period.new years=10) > (Period.new days=1)) . should_fail_with Incomparable_Values
+            ((Period.new years=10 months=3) > (Period.new months=5)) . should_fail_with Incomparable_Values
```
Comment (radeusgd) [inline test/Tests/src/Data/Time/Period_Spec.enso:34]: It's constantly confusing me that this is a 0-constructor 'singleton' type.

By the way - wouldn't it make more sense to extend this error to contain in its payload the two offending values which were unable to be compared? I think it may be easier when debugging to immediately see what kind of objects were attempted to be compared and failed.

File: test/Tests/src/Network/Http/Request_Spec.enso [hunk 20]
```diff
@@ -13,7 +14,7 @@ spec =
     test_headers = [Header.application_json, Header.new "X-Foo-Id" "0123456789"]
     Test.group "Request" <|
         Test.specify "should return error when creating request from invalid URI" <|
-            Request.new HTTP_Method.Post "invalid uri" . should_fail_with Syntax_Error_Data
+            Request.new HTTP_Method.Post "invalid uri" . should_fail_with Syntax_Error.Error
```
Comment (radeusgd) [inline test/Tests/src/Network/Http/Request_Spec.enso:17]: Nothing new here, but looking at this test I'm not so certain that having one `Syntax_Error` for all kinds of errors is a good idea - a URI parse error is a different thing from a Regex syntax error from ... etc.

Not necessarily to address in this PR, but maybe worth considering - I think we may want to catch a specific kind of error only, not all syntax errors regardless of their origin?

**Review by radeusgd — APPROVED** [review]

**Comment by radeusgd** [comment]
image

**Comment by hubertp** [comment]
> image

image
